### PR TITLE
Refactor Find Sources and fix bug when view a user who belongs to an unactive auth source

### PR DESCRIPTION
**Pull request by lunny**
The steps to reproduce it.

First, create a new oauth2 source.
Then, a user login with this oauth2 source.
Disable the oauth2 source.
Visit users -> settings -> security, 500 will be displayed.
This is because this page only load active Oauth2 sources but not all Oauth2 sources.

### CODE REVIEW

**Review by nanguanlin6 — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

File: routers/web/user/setting/security/security.go [hunk 1]
```diff
@@ -105,11 +107,31 @@ func loadSecurityData(ctx *context.Context) {
...
+			ctx.ServerError("CreateProviderFromSource", err)
+			return
+		}
+		oauth2Providers[source.Name] = provider
+		if source.IsActive {
+			orderedOAuth2Names = append(orderedOAuth2Names, source.Name)
+		}
+	}
+
+	sort.Strings(orderedOAuth2Names)
```
Comment (KN4CK3R) [inline routers/web/user/setting/security/security.go:133]: We should replace this with
```go
sort.Slice(authSources, func(i, j int) bool { return authSources[i].Name < authSources[j].Name })
```
and just use the sorted slice in the template. I think this code was pre Go 1.8 when there was no easy way to sort a slice of structs.

Comment (lunny) [reply routers/web/user/setting/security/security.go:133]: Done with other places, but for this place. We have to use an active name slice and an all auth sources to handle it.

**Review by techknowlogick — COMMENTED** [review]
There are some changes to these functions in https://github.com/go-gitea/gitea/pull/25165 to make them more generic. Could you utilize some of these update function names/args to make conflict merging easier?

**Comment by lunny** [comment]
> There are some changes to these functions in #25165 to make them more generic. Could you utilize some of these update function names/args to make conflict merging easier?

I think the changes are the correct direction. I will help to resolve your PR's conflict after this merge.

**Comment by lunny** [comment]
There are too many conflicts, so I think maybe it's not necessary to backport to v1.21 and v1.20.

**Pull request merged**